Source/bmalloc/libpas/src/libpas/pas_stats.h

```diff
@@ -0,0 +1,311 @@
+ *       allocate memory of its own, but instead prefer calling
+ *       `pas_stats_ensure_print_buffer` with whatever size of memory is
+ *       necessary.
+ *   2. Define a new PAS_RECORD_STAT_<statname> macro with the desired arguments
+ *     - this should call the function which actually modifies your stat-struct;
+ *       this function must handle its own synchronization.
+ *   3. Add calls to PAS_RECORD_STAT(<statname>, <args...>) to the relevant
+ *      points inside of libpas.
+ * The fact that we need #2 is somewhat undesirable (in principle we should just
+ * generate that table automatically from PAS_STATS_FOR_EACH_COUNTER) but I
+ * couldn't find a way to do so ergonomically.
+ *
+ * Since this framework is intended for use inside of an allocator, it is
+ * intended to have low overhead (both for enabled and disabled counters) and to
+ * have minimal use of heap-allocated memory -- however, there is room for
+ * improvement on both counts.
+ *   - Re.: heap-allocated memory, on the logging path we currently do rely on
+ *     heap allocations to make it easier for people to add new counters, as
+ *     using a fixed-size static allocation per counter would mean every counter
+ *     would need to pre-compute the theoretical maximum size of its json
+ *     payload.
+ *     Normally the utility heap would be a good fit for this use-case, but to
+ *     avoid reentrancy we do not use libpas to allocate this memory -- even by
+ *     going through the system heap. Instead, we call `malloc` directly. These
+ *     buffers are cached so it shouldn't happen often but it would be better to
+ *     be able to remove that dependency.
+ *   - Re.: performance, the current design is not bad but does introduce a lot
+ *     of atomic traffic and cross-core contention. Ideally, we would instead
+ *     have a per-thread 'local stat counter cache' which we would then
+ *     periodically accumulate into a global stat-counter object. Individual
+ *     stat counters would have to be aware since they need to implement their
+ *     own accumulate functions.
+ *     Even better than thread-local would be if we had something like Linux'
+ *     rseqs, as we could then store this data per-CPU and avoid any migration
+ *     whatsoever.
+ *     In both cases though, we would risk under-counting statistics unless we
+ *     implemented an analog of what pas-TLCs do where they iterate over other
+ *     threads' TLCs and collect data out. Doing so generically across all kinds
+ *     of stat counters seems like a challenge.
+ */
+
+#include "pas_config.h"
+
+#include "pas_allocation_mode.h"
+#include "pas_lock.h"
+#include "pas_internal_config.h"
+#include "pas_utils.h"
+#include <stdbool.h>
+#include <stdatomic.h>
+#include <stdint.h>
+#include <stdio.h>
+
+PAS_BEGIN_EXTERN_C;
+
+/**********************************/
+/* General stat-counter machinery */
+/**********************************/
+
+typedef char* (*pas_stats_stat_logger)(void* stat);
```

Comment:
nit: let's call this `pas_stats_stat_logger_function`.  I think that helps makes it clear that this is a function pointer and not a struct when used as a fields as is the case in the definition of `pas_stats_stat_base` below.

You can also shorten it to `pas_stats_logger_function`.  I'm not sure the extra `stat` adds much context.